### PR TITLE
Download integrity follow-ups: audit, artifact sweep, MissAV source duration

### PR DESCRIPTION
**Pull request by franklioxygen**
Implements items 4, 5, 6 and 8 from `reports/download-integrity-followups-design-2026-09-21.md`, the open items after [#471](https://github.com/franklioxygen/MyTube/pull/471). Item 7 needs a decision and is not here — see the end.

Four commits, one per item, in the order the design recommended.

## 8 — Reject a MissAV run that produced no output

The completeness check is deliberately fail-open: a file it cannot read is reported as unknown, not broken, so a host without ffprobe keeps working. That means it cannot answer *"yt-dlp produced nothing at all"*.

Bilibili and the yt-dlp path gate on that separately, through `findVideoFileInTemp` and `resolvePlayableMediaFilePath`. MissAV did not — its only existence test was for file size, and when that failed it left the size undefined and saved the row anyway. Pre-existing, and not something #471 covered, though its presence could be mistaken for covering it.

Gates on existence before probing. The probe keeps its fail-open contract: the existence question belongs to the caller that knows which path it asked for.

Also pins the other two downloaders' gates with a test each — neither was covered.

## 5 — Library-wide media integrity audit

`GET /api/media-integrity-audit` (admin), following the shape of the existing media-collision audit. Reports per row: `file_missing`, `track_disagreement`, `duration_mismatch`, `unprobeable`.

Reuses `probeMediaTrackDurations` and `evaluateMediaCompleteness` rather than reimplementing either, and the shortfall tolerance is now exported as `allowedDurationDrift` and shared, so the audit and the download-time check cannot drift apart about what "truncated" means.

**Read-only by design.** It recommends, never repairs: re-downloading is expensive, can fail, and for a source that has gone offline would replace a partially watchable file with nothing.

Two details that matter on a real library: `cloud:`, `mount:` and `http(s)` rows are skipped rather than counted as missing — an ad-hoc version of this check reported two false positives for exactly that reason — and probe results are cached by mtime+size.

## 4 — Detect abandoned download artifacts

`POST /api/cleanup-temp-files` covers `.part`/`.ytdl` files and temp directories carrying the ownership marker. Two kinds escape it:

- **Unmarked temp directories** — the marker was added *after* the temp directory mechanism, so anything older is invisible to `isOwnedInactiveDownloadTempDir` and can never be cleaned however often the endpoint runs.
- **Split-stream artifacts** — `*.fNNN.*` / `*.temp.*`, written straight into the library by the non-Bilibili downloaders.

One production library held 565 MB across ten of them, including a 136 MB directory that had to be deleted by hand.

**Deletion is implemented and tested but armed nowhere.** The startup job reports and removes nothing. These artifacts run to hundreds of megabytes, clearing them unprompted is not a restart's decision, and arming it needs the cleanup dialog's copy updated — it currently discloses only `.part` and `.ytdl` removal.

Every candidate must clear three independent conditions — known artifact name shape, older than seven days, referenced by no library row — so one check misfiring cannot select a file still in use. A user folder named `temp_holidays` does not match the directory pattern, which is the shape of the earlier P1.

> Writing the tests caught a real defect in the first draft: it descended into download temp directories and collected their contents as individual candidates, which for an **active** download meant selecting the `.part` file of a transfer in progress. Temp directories are now never descended into.

## 6 — MissAV source duration from the HLS playlist

MissAV was the one path passing `null`, leaving it with only the track comparison and blind to a download shortened equally across both tracks.

The playlist is a better source than the page: it describes the exact stream being fetched, so it cannot disagree with the rendition yt-dlp downloads. A master playlist is followed one level to its first variant — every variant of one VOD is the same content at a different bitrate, so which one answers is irrelevant, and the master is what the URL selector usually picks.

`EXT-X-ENDLIST` is required: without it the stream is live or still being written and the sum would undercount, which is the one direction that rejects a good download. Every other uncertain case returns null — the behaviour before this — so the resolver can only add detection.

Also mocks axios in the MissAV suite. It was unmocked, so adding this fetch made the tests issue real outbound requests; they passed only because those requests failed.

## Testing

- `mediaIntegrityAuditService.test.ts` — 16 cases
- `downloadArtifactSweep.test.ts` — 19 cases, including the armed deletion path and the active-download guard
- `missavM3u8Duration.test.ts` — 11 cases, mostly the null paths
- MissAV suite 62, downloaders suite unchanged otherwise
- `tsc --noEmit` clean; full backend suite **3,463 passed**

Six failures in `ytdlpHelpers` ×4, `favoriteMigration`, `sourceVideoIdMigration` and `coreDataSafety.integration` are **pre-existing** — verified failing on a clean baseline.

## Item 7 is not here

`--abort-on-unavailable-fragments` for MissAV HLS is a genuine trade-off, not an oversight: a gap-free file or an honest failure, versus a single transient CDN 404 failing a long download that would otherwise have completed with a fraction of a second missing.

**Item 6 weakens the case for it.** The reason aborting looked attractive was that dropped fragments shorten both tracks together and evade the track comparison — which is exactly what a working source duration now catches. Worth deciding deliberately rather than bundling here.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added an admin-only, read-only media integrity audit for missing, unreadable, temporary, or inconsistent local files.
  - Added startup reporting for unreferenced download artifacts older than seven days, without automatic deletion.
  - Added MissAV HLS duration validation using browser-captured playlists, proxy-aware fallback requests, and rendition comparison.

- **Bug Fixes**
  - Downloads now fail when no output file is produced.
  - Improved media completeness checks using source playlist duration when available.
  - Invalid stored durations are now rejected instead of partially parsed.

### CODE REVIEW

File: backend/src/services/downloaders/MissAVDownloader.ts [hunk 1]
```diff
@@ -620,6 +620,33 @@ export class MissAVDownloader extends BaseDownloader {
...
+      // case yields null, which is the behaviour without this.
+      const sourceDurationSeconds = await resolveM3u8DurationSeconds(
+        m3u8Url,
+        async (target) => {
+          const axios = (await import("axios")).default;
+          const response = await axios.get(target, {
+            ...(typeof userConfig.proxy === "string"
+              ? getAxiosProxyConfig(userConfig.proxy)
+              : {}),
+            headers: { Referer: referer },
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/downloaders/MissAVDownloader.ts]: Fixed in bd573a7b — confirmed, and the consequence was worse than a failed fetch.

The impersonation note sits a few lines below that request and says the global `--impersonate` flag is required because even yt-dlp's `extractor-args generic:impersonate` leaves the manifest fetch on the default fingerprint and gets a 403, *verified against surrit.com from the deployment environment*. A plain Axios request was never going to clear that.

And because `resolveM3u8DurationSeconds` swallows the failure, it was **silent**: the source duration would simply never be established on the preferred host, the check would fall back to comparing tracks, and a download shortened equally across both tracks would go undetected — the exact gap this was added to close.

Took the second of your two suggestions. The Puppeteer step that extracts the m3u8 URL has already fetched these playlists with an accepted fingerprint and the session cookies, so their bodies are now captured in the existing `page.on('response')` handler and reused:

- no extra request, and it is the only thing that works on a host that fingerprints
- bodies are read before `browser.close()`, since a response cannot be resolved afterwards
- a body that cannot be read (redirect, aborted request) just leaves that URL uncaptured
- a URL the browser never fetched — a variant it did not select — is still requested directly, which works on hosts that do not fingerprint and otherwise fails into an unknown duration, i.e. the behaviour without this lookup at all

I did not route it through yt-dlp `--impersonate`: that would mean an extra process spawn to re-fetch something the browser already has in hand.

Covered by a test asserting the captured body produces the duration **and** that axios is not consulted for that URL, so a regression back to the direct fetch fails the suite rather than going quiet again.

File: backend/src/services/mediaIntegrityAuditService.ts [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+  };
+  const items: MediaIntegrityAuditItem[] = [];
+
+  type Candidate = { video: Video; absolutePath: string };
+  const candidates: Candidate[] = [];
+  let collections: ReturnType<typeof storageService.getCollections> | undefined;
+
+  for (const video of videos) {
+    const webPath = readString(video.videoPath);
+    if (webPath && /^(?:cloud:|mount:|https?:\/\/)/.test(webPath)) {
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/mediaIntegrityAuditService.ts]: Fixed in bc4676de. Applied your diff as proposed — the premise checks out: `resolveManagedWebPath` matches lowercase schemes only (pathHelpers.ts:75-76), so an uppercase-scheme row fell through to the local branch, resolved to null, and was reported as a missing file.

Covered by a parameterised test over `HTTP://`, `HTTPS://`, `Cloud:` and `MOUNT:` asserting each is counted as external with no finding raised.

File: backend/src/services/mediaIntegrityAuditService.ts [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+  try {
+    stat = statSafeSync(absolutePath, VIDEOS_DIR);
+  } catch {
+    stat = null;
+  }
+
+  if (stat) {
+    const cached = probeCache.get(absolutePath);
+    if (cached && cached.mtimeMs === stat.mtimeMs && cached.size === stat.size) {
+      return cached.tracks;
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/mediaIntegrityAuditService.ts]: Fixed in bc4676de, though not the way you suggested — I took the bounded-age option rather than a content digest.

The concern is real: mtime and size catch every replacement this application performs (a re-download or repair writes a new file, so both change), but they are not proof of identity. A file rewritten in place to the same length with its timestamp restored would read as unchanged, and the audit would keep reporting the stale result until the process restarted.

I did not use a content digest because it inverts the point of the cache: the cache exists to avoid an ffprobe call, and hashing the whole media file to validate an entry costs substantially more than the probe it is saving. On a large library that would make a repeat audit slower than having no cache at all.

Bounded the entry age instead. A repeat audit in the same sitting is still cheap, staleness is capped at ten minutes, and mtime+size continue to invalidate immediately in every case the application itself causes — so the only window is a hand-crafted in-place rewrite, which is now ten minutes rather than until restart.

Covered by a test using fake timers: an entry with unchanged mtime and size is re-probed once it ages past the maximum.

File: backend/src/services/downloaders/MissAVDownloader.ts [hunk 4]
```diff
@@ -620,6 +651,39 @@ export class MissAVDownloader extends BaseDownloader {
...
+        m3u8Url,
+        async (target) => {
+          const captured = capturedPlaylists.get(target);
+          if (captured !== undefined) return captured;
+          // Not something the browser fetched - a variant it never selected,
+          // say. Worth trying directly: it succeeds on hosts that do not
+          // fingerprint, and on those that do it fails into an unknown
+          // duration, which is the behaviour without this lookup at all.
+          const axios = (await import("axios")).default;
+          const response = await axios.get(target, {
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/downloaders/MissAVDownloader.ts]: Fixed in 32ba0441 — correct, and it is a genuine interaction between the two changes rather than a leftover: corroboration created a second read that the captured-body fix does not cover.

I did not route the fallback through an impersonating client. Doing it properly would mean either keeping the browser open across the URL selection — which happens after `browser.close()`, and includes a regex-reconstruction fallback that can produce a URL the browser never fetched — or relaunching Chromium for a single manifest read. Both are a lot of moving parts for one number, and neither helps the case where the CDN simply never served that rendition to anyone.

Two smaller changes instead:

**Read what is already in hand first.** The captured URLs are passed down and renditions present in that set are ordered ahead of the rest, so the one the browser loaded falls inside the two-rendition budget instead of sitting behind one that cannot be read.

**Stop letting an unreadable rendition discard a readable one.** A failed read is now skipped rather than aborting. With only one rendition in hand there is nothing to corroborate against, so it is used — which is what this did before corroboration was added, and it beats reporting no duration at all. The completeness check tolerates a shortfall of `max(10s, 5%)`, far more than renditions of one VOD realistically differ by. Corroboration still applies wherever two can be read, and two that disagree are still refused outright.

So the ordering matters for detection and the fallback matters for coverage; neither weakens the disagreement guard.

Three tests: one readable rendition yields the duration, no readable rendition still yields null, and the ordering hint is honoured (without it the unreadable rendition is tried first and nothing is left within budget).

File: backend/src/services/mediaIntegrityAuditService.ts [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+      summary.unprobeable += 1;
+    } else {
+      // No live source lookup here, so only the track comparison applies - the
+      // same position MissAV downloads are in.
+      if (!evaluateMediaCompleteness(tracks, null).complete) {
+        reasons.push("track_disagreement");
+        summary.trackDisagreements += 1;
+      }
+
+      const stored = parseSourceDurationSeconds(video.duration);
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/mediaIntegrityAuditService.ts]: Fixed in 32ba0441.

Worth noting the premise is narrower than stated: every writer in this codebase stores whole seconds, and `parseSearchDuration` exists precisely to convert Bilibili's colon notation *before* it reaches storage. So this is not a format the application produces. But the column is free-form text, `formatRssDuration` already passes colon notation straight through — which is evidence the codebase treats it as possible — and a restored or hand-edited row can carry anything.

The consequence you describe is right: `parseFloat("01:23")` is 1, so an intact 83-second file would be reported as a `duration_mismatch` with a misleading refresh recommendation.

Took both of your options, split by whether the value is unambiguous. Clock notation is parsed, since `HH:MM:SS` and `MM:SS` have exactly one reading. Anything else yields no duration, which skips the comparison — the audit should not manufacture a finding out of a value it does not understand.

Scoped to the audit rather than changing `parseSourceDurationSeconds`, which is shared with the download-time check where the value always comes from ffprobe or source metadata and is numeric by construction.

Tests cover `01:23`, `1:23` and `01:02:03` against matching files, an uninterpretable value being skipped, and a clock-formatted duration that genuinely disagrees still being flagged with the correctly parsed seconds.

File: backend/src/services/downloaders/MissAVDownloader.ts [hunk 6]
```diff
@@ -620,6 +651,40 @@ export class MissAVDownloader extends BaseDownloader {
...
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/downloaders/MissAVDownloader.ts]: Fixed in 4cfa0881 — agreed, and this is the more serious of the two findings on this file.

The playlist body is attacker-controllable, so an absolute variant URI turned straight into a backend GET wherever it pointed. My own test even exercised a cross-origin absolute URI, so this was working as written rather than an edge case.

Restricted renditions to the master's own origin — where a rendition of it belongs — or to a URL the browser already fetched under its own policy during playback. Anything else is skipped rather than requested, so the resolver simply reports no duration.

Chose origin matching over the existing `validateUrl` allowlist helper: that blocks private and internal addresses, but a public attacker-controlled host would still be fetched, and same-origin needs no allowlist to keep current.

Tests assert `127.0.0.1`, `169.254.169.254` and an unrelated public host are never requested — the fake fetcher throws if they are, so a regression fails loudly — plus one asserting an off-origin rendition the browser did fetch is still allowed.

File: backend/src/services/downloaders/missav/m3u8.ts [hunk 7]
```diff
@@ -103,3 +108,172 @@ export function selectBestM3u8Url(
...
+
+  // Read renditions already in hand first. On a CDN that fingerprints TLS, a
+  // rendition the browser never loaded is answered with 403, so ordering decides
+  // whether anything is readable at all.
+  if (readableUrls?.size) {
+    urls.sort((a, b) => Number(readableUrls.has(b)) - Number(readableUrls.has(a)));
+  }
+
+  const durations: number[] = [];
+  for (const variantUrl of urls.slice(0, MAX_COMPARED_VARIANTS)) {
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/downloaders/missav/m3u8.ts]: Fixed in 4cfa0881. Correct, and it is the same failure this change set has now hit twice: anything that makes the resolver return null silently disables source-duration validation.

`slice(0, MAX_COMPARED_VARIANTS)` chose the candidates before any were read, so two failures exhausted the budget even with readable renditions behind them. Iteration now continues until two durations are collected.

Added a separate attempt bound rather than leaving it open: on a protected host every unreadable rendition costs a round trip, so a master advertising a dozen of them would otherwise mean a dozen 403s for one number.

Two tests: one where the first two renditions fail and later ones succeed, and one asserting a master of twelve unreadable renditions stops after a bounded number of attempts.

File: backend/src/services/mediaIntegrityAuditService.ts [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+      const filename = readString(video.videoFilename);
+      // Older rows identify their file by filename and collection only. Missing
+      // or invalid local paths are findings, not evidence of an external video.
+      const absolutePath = webPath
+        ? (resolved?.prefix === "/videos" ? resolved.absolutePath : null)
+        : (filename ? storageService.findVideoFile(filename, (collections ??= storageService.getCollections())) : null);
+
+      if (TEMPORARY_VIDEO_ARTIFACT_PATTERN.test(path.basename(absolutePath ?? webPath ?? filename ?? ""))) {
+        summary.skippedTemporaryArtifacts += 1;
+        continue;
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/mediaIntegrityAuditService.ts]: Fixed in 4cfa0881 — you are right, and the skip is gone entirely rather than being made smarter.

The pattern is the codebase's own definition of a yt-dlp intermediate, and it is correct where it came from: the metadata backfills use it to skip loose files found by scanning the directory. But the audit iterates database rows, and a row states that this file *is* its published media. Judging that by basename inverts the tool's purpose — it would hide corruption in exactly the entries it exists to check.

I did not replace it with a record-state check, because there is no state that distinguishes 'row pointing at an intermediate' from 'row with an unusual filename', and a row pointing at a genuine intermediate is itself a finding worth reporting rather than suppressing. So every row the database claims is published now gets audited, and the `skippedTemporaryArtifacts` summary field is gone with it.

Tests cover `a.f137.mp4`, `lesson.f137.final.mp4` and `a.temp.mp4`, each asserting the file is probed and its corruption reported.

File: backend/src/services/mediaIntegrityAuditService.ts [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+ */
+function parseStoredDurationSeconds(value: unknown): number | null {
+  if (typeof value === "string" && value.includes(":")) {
+    const parts = value.trim().split(":");
+    if (parts.length < 2 || parts.length > 3) return null;
+    if (!parts.every((part) => /^\d+$/.test(part))) return null;
+    const seconds = parts.reduce((total, part) => total * 60 + Number(part), 0);
+    return Number.isFinite(seconds) && seconds > 0 ? seconds : null;
+  }
+  return parseSourceDurationSeconds(value);
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/mediaIntegrityAuditService.ts:172]: Fixed in 4cfa0881. Same class as the clock-format issue in the previous round, and my fix for that one left the hole open: the non-clock branch still delegated to `parseSourceDurationSeconds`, which is parseFloat-based.

That leniency is right where it came from — ffprobe output and source metadata are numeric by construction — but wrong for a free-form column, where `"1 hour"` reads as 1 and a 3600-second file is reported as mismatched by 3599 with a refresh recommendation.

Non-clock values must now match an exact numeric representation; anything else yields no duration and skips the comparison. Scoped to the audit, so `parseSourceDurationSeconds` keeps its existing behaviour for the download path.

Tests cover `1 hour` as you suggested, plus `12abc`, empty, whitespace and `NaN`.

File: CHANGELOG.md [hunk 10]
```diff
@@ -4,10 +4,18 @@
...
...osts that do not fingerprint and otherwise fails into an unknown duration. The playlist is a better source than the page: it describes the exact stream being fetched, so it cannot disagree with the rendition yt-dlp downloads, whereas a page-scraped number describes the title. A master playlist is corroborated rather than guessed at: yt-dlp, not this code, selects the rendition that gets downloaded, so two are read and compared, and the master is unusable when they disagree by more than the completeness check itself tolerates - which is exactly the condition under which using either one cannot by itself cause a rejection. The shorter of the two is taken, since understating the source is the direction that cannot produce a false rejection. A master offering separate audio or subtitle renditions is refused outright, because each variant then covers only part of the muxed result. `EXT-X-ENDLIST` is required because a live or growing playlist cannot establish the final duration. Malformed playlists and failed requests also yield no duration, preserving the track comparison as the fallback. The playlist request honors the download's proxy setting, including an explicit direct connection.
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply CHANGELOG.md]: Fixed in 4cfa0881. Correct — the text described the implementation as it stood one commit earlier, before the single-readable fallback was added, so it understated coverage.

Rewritten to say up to two renditions are read; two that disagree beyond the tolerance make the master unusable; otherwise the shorter is taken; and when only one can be read it is used as it stands, with the reason stated (a CDN that 403s anything the browser did not fetch, which is the usual case). Also documented the origin restriction added for the SSRF finding on the same file, since that bounds which renditions are eligible at all.

File: backend/src/services/mediaIntegrityAuditService.ts [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+function resolveAction(
+  reasons: MediaIntegrityAuditReason[]
+): MediaIntegrityRecommendedAction {
+  if (reasons.includes("file_missing") || reasons.includes("track_disagreement")) {
+    return "redownload";
+  }
+  if (reasons.includes("unprobeable")) {
+    return "manual_review";
+  }
+  return "refresh_duration";
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/mediaIntegrityAuditService.ts:230]: Fixed in 938c2683.

Agreed, and the part that makes it worth fixing rather than noting is that `refresh_duration` is specifically the action that destroys the evidence. Overwriting 3600 with 1800 removes the only record that the file used to be longer, so every later audit passes and the corruption becomes permanently invisible — in a tool whose entire purpose is to surface exactly that.

The direction carries the meaning, so the action now follows it: longer than stored is stale metadata and still refreshes; shorter than stored has the shape of a truncation after the row was written and recommends a redownload. The detail says so explicitly rather than leaving a reader to compare the two numbers themselves.

Kept it as one `duration_mismatch` reason with a direction-aware action rather than splitting the reason, since the numbers and the explanation are both already in the finding.

Test asserts a 3600-second row measuring 1800 reports `redownload`, alongside the existing longer-than-stored case still reporting `refresh_duration`.

File: backend/src/services/downloaders/MissAVDownloader.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/downloaders/MissAVDownloader.ts]: Fixed in f8393443. Correct — the previous fix validates the origin *before* the request, and Axios follows redirects by default, so a same-origin rendition answering 302 walks straight past it.

Refused redirects outright (`maxRedirects: 0`) rather than revalidating each hop. Validating hops means a `beforeRedirect` callback, which ties this to the shape of follow-redirects' options object — more surface to get subtly wrong for a case that buys little: a rendition referenced from a master on its own origin has no particular reason to redirect.

The degradation is the same as everywhere else here. A 3xx fails Axios' default status validation, the throw is caught per-rendition, that rendition is skipped, and if nothing else is readable the result is no duration — which is the behaviour without this lookup at all. It also does not touch the primary path on protected hosts, where the body comes from what the browser already fetched and no request is made.

Scoped to this fetch. The thumbnail download in the same file goes through the base class and takes its URL from the page's og:image rather than a playlist, so it is a different exposure and I have not changed it here.

Test asserts the request carries `maxRedirects: 0`.

File: backend/src/services/downloaders/MissAVDownloader.ts [hunk 12]
```diff
@@ -414,6 +442,9 @@ export class MissAVDownloader extends BaseDownloader {
         }
 
         html = await page.content();
+        // The bodies must be read while the page is still alive; after
+        // browser.close() the responses can no longer be resolved.
+        await Promise.allSettled(pendingCaptures);
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/downloaders/MissAVDownloader.ts]: Fixed in c6a776eb. Confirmed, and this one is mine — introduced with the body capture in the same round.

The consequence is worse than losing the duration, which is what makes it worth more than a tweak: the wait sits in front of `browser.close()` and the yt-dlp spawn, and `onStart` has not registered the cancellation callback yet, so a single stalled response hangs the download with no way to abort it. Trading a missing duration for an unkillable download is a bad deal in both directions.

Bounded the aggregate wait rather than each capture: `allSettled` already waits in parallel, so one cap bounds the total regardless of how many are outstanding. Anything still pending is abandoned and the download proceeds without it — falling back to reading that playlist directly, or to no duration at all, which is where this started.

Abandoned rather than cancelled, so a late body can still resolve after the fact; it either writes into a map nobody reads again or rejects because the page is gone, and both are harmless.

Put the helper in `utils/concurrency` beside `runWithConcurrencyLimit` instead of inlining it, since the bound is the entire point and worth testing directly — including that it clears its timer on the happy path, because a lingering one would keep the process alive.

File: backend/src/services/downloaders/MissAVDownloader.ts [hunk 13]
```diff
@@ -620,6 +661,50 @@ export class MissAVDownloader extends BaseDownloader {
...
+      // MissAV is the one path with no source duration, which leaves it with
+      // only the track comparison and blind to a download shortened equally
+      // across both tracks. The playlist answers it authoritatively: it
+      // describes the exact stream being fetched. Best-effort - every uncertain
+      // case yields null, which is the behaviour without this.
+      const sourceDurationSeconds = await resolveM3u8DurationSeconds(
+        m3u8Url,
+        async (target) => {
+          const captured = capturedPlaylists.get(target);
+          if (captured !== undefined) return captured;
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/downloaders/MissAVDownloader.ts]: Fixed in 0dfae01c. Confirmed — and it is the two previous fixes combining into a third failure, which is worth naming: the request listener records the pre-redirect URL, so that is what the selector picks; the body arrives under the post-redirect URL; and the direct fallback now refuses redirects, so the miss produced no duration at all rather than falling back to a request.

Bodies are now keyed under every hop in the chain (`response.request().redirectChain()`) alongside the final URL, so a lookup for either end finds it. Keying rather than retrying, as you suggested — retrying the original URL directly would just hit the redirect refusal again.

Test models a response arriving under a post-redirect URL with the pre-redirect URL in its chain, and asserts the duration is produced **and** that axios is never consulted.

File: backend/src/services/downloaders/MissAVDownloader.ts [hunk 14]
```diff
@@ -358,6 +376,21 @@ export class MissAVDownloader extends BaseDownloader {
...
+
+          if (response.status() !== 200 || capturedPlaylists.has(resUrl)) return;
+          // Best-effort: a body that cannot be read (redirect, aborted request,
+          // page already gone) simply leaves this URL uncaptured, and the
+          // duration lookup falls back to requesting it directly.
+          pendingCaptures.push(
+            response
+              .text()
+              .then((body) => {
+                if (typeof body === "string" && body.length <= MISSAV_MAX_CAPTURED_PLAYLIST_BYTES) {
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/downloaders/MissAVDownloader.ts]: Fixed in 0dfae01c. Right on both counts — the guard discarded rather than prevented, and nothing bounded the aggregate at all.

Three changes: a body whose declared `content-length` already exceeds the per-body cap is no longer read, so it is not allocated in order to be thrown away; the number of retained bodies is capped; and their total size is capped.

One detail worth flagging: the count is tracked separately from the map rather than using `map.size`, because the redirect fix in the sibling comment makes the map hold one entry per hop. Using its size would overstate how many bodies are actually retained and would tighten the cap unpredictably depending on how many redirects the CDN happens to use.

The declared-length check is not a complete guarantee, since `content-length` may be absent on a chunked response — the post-read length check still backs it up, and the aggregate cap bounds the rest.

File: backend/src/services/downloaders/MissAVDownloader.ts [hunk 15]
```diff
@@ -358,6 +385,56 @@ export class MissAVDownloader extends BaseDownloader {
...
+                // A redirected request is recorded by the request listener under
+                // the URL it started at, while the body arrives under the URL it
+                // ended at. Key it under every hop so a lookup for either finds
+                // it - otherwise the selected URL misses, and the direct fallback
+                // refuses the redirect, leaving no duration at all.
+                for (const hopUrl of [
+                  resUrl,
+                  ...response.request().redirectChain().map((hop) => hop.url()),
+                ]) {
+                  capturedPlaylists.set(hopUrl, body);
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/downloaders/MissAVDownloader.ts:423]: Fixed in 02df32b1. Correct, and it follows directly from the redirect-aliasing fix in the previous round: making the body reachable under the pre-redirect URL also made that the URL the resolver treats as the master's location, since it is what the request listener recorded and therefore what the selector picks.

HLS resolves relative URIs against the final location, so the resolver now takes the URL the body actually arrived from as its base. The aliases stay usable for lookup — that part was right — but they no longer decide where the variants live.

Passed as an optional `finalUrlOf` hook rather than changing the fetch contract to return a URL alongside the body: only the master's base needs it, and the hook leaves every existing caller and test untouched.

Test covers a master whose body is fetched under `/original/playlist.m3u8` but arrives from `/final/playlist.m3u8`, with variants that only resolve correctly against the latter.

File: backend/src/services/downloaders/MissAVDownloader.ts [hunk 16]
```diff
@@ -358,6 +385,56 @@ export class MissAVDownloader extends BaseDownloader {
               `server=${headers["server"] ?? "?"} ` +
               `set-cookie=${headers["set-cookie"] ? "yes" : "no"} ${resUrl}`,
           );
+
+          if (response.status() !== 200 || capturedPlaylists.has(resUrl)) return;
+          if (
+            capturedPlaylistCount >= MISSAV_MAX_CAPTURED_PLAYLISTS ||
+            capturedPlaylistBytes >= MISSAV_MAX_CAPTURED_PLAYLIST_TOTAL_BYTES
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/downloaders/MissAVDownloader.ts]: Fixed in 02df32b1. Right — the caps bounded what was *kept*, not what was *read*, which is where the memory actually goes.

Slots are now reserved before the read starts and released if it fails or the body turns out unusable. A response with no `content-length` reserves the per-body cap and reconciles once the real size is known, so an undeclared body cannot slip past the aggregate limit by being unmeasurable up front.

This also bounds `pendingCaptures` itself, since a promise is only pushed after a slot is secured.

The test fires 80 concurrent matching responses whose bodies never resolve, and asserts `text()` was called no more than the cap — bodies that never settle keep every slot occupied, so the count cannot be satisfied by reads completing and freeing slots. It doubles as a check that the capture wait stays bounded, since the download still completes with none of them resolving.

File: backend/src/services/downloaders/MissAVDownloader.ts [hunk 17]
```diff
@@ -358,6 +388,83 @@ export class MissAVDownloader extends BaseDownloader {
...
+              .text()
+              .then((body) => {
+                if (
+                  typeof body !== "string" ||
+                  body.length > MISSAV_MAX_CAPTURED_PLAYLIST_BYTES
+                ) {
+                  settleSlot(null);
+                  return;
+                }
+                settleSlot(body.length);
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/downloaders/MissAVDownloader.ts]: Fixed in 3ca4adc2. Correct — the reservation was the only check on the aggregate, and a compressed transfer's `Content-Length` can be a fraction of the decoded body, so concurrent responses could each reserve a little, pass, and then be kept at nearly 4 MiB apiece.

The decoded size is now checked against both the per-body and aggregate limits before a body is kept; a body that no longer fits releases its room instead.

I also moved this bookkeeping out of the response handler. It has needed a fix in each of the last three review rounds (count after read, then reservation vs. aggregate, now this), and inside the closure none of it was observable to a test. It is now `PlaylistCaptureBudget` in `missav/captureBudget.ts`, where a reservation returns a single-use ticket, and it is tested directly — including your exact case: 32 responses each declaring 1 KiB and decoding to just under 4 MiB, of which only 4 are kept.

File: backend/src/services/mediaIntegrityAuditService.ts [hunk 18]
```diff
@@ -0,0 +1,392 @@
...
+    }
+
+    try {
+      const resolved = webPath ? resolveManagedWebPath(webPath) : null;
+      const filename = readString(video.videoFilename);
+      // Older rows identify their file by filename and collection only. Missing
+      // or invalid local paths are findings, not evidence of an external video.
+      const absolutePath = webPath
+        ? (resolved?.prefix === "/videos" ? resolved.absolutePath : null)
+        : (filename ? storageService.findVideoFile(filename, (collections ??= storageService.getCollections())) : null);
```
[the remark this answers is not in the document]
Comment (franklioxygen) [reply backend/src/services/mediaIntegrityAuditService.ts:299]: Fixed in 3ca4adc2. Confirmed — `findFileInStorage` catches, logs and returns null — and the problem was wider than the legacy path.

Rows with a `videoPath` go through `pathExistsSafeSync`, which is `fs.existsSync` underneath and answers false for `EACCES` or `EIO` exactly as for a missing file. So a dropped NAS mount would have reported the whole library missing and recommended redownloading all of it, not only the legacy rows.

Both paths now confirm a miss before reporting it:

- **With a path:** `stat` it, and only `ENOENT` or `ENOTDIR` count as absent. Anything else is thrown into the per-row catch and the row is reported `unprobeable` with the error.
- **Legacy lookup miss:** there is no path to stat, so the library directory must be readable before the row is called missing. Checked once per audit, with the result reused.

One limit I did not paper over: an unmounted mountpoint is an empty, readable directory, so every file in it really does stat as `ENOENT`. No per-file check can tell that apart from files that are genuinely gone. That would need a library-level heuristic — something like "most rows missing at once" — which I have not added here.

Tests cover `ENOENT` still reporting missing, `EACCES`/`EIO`/`ESTALE` reporting unprobeable, the legacy miss with the library readable and unreadable, and the library check running once rather than per row.
